packages/sitecore-jss/src/personalization/layout-personalization-utils.ts

```diff
@@ -0,0 +1,123 @@
+import {
+  PersonalizedComponentRendering,
+  isComponentRendering,
+  isPersonalizedComponentRendering,
+  ComponentRendering,
+  PlaceholdersData,
+  hasPersonalization,
+  HtmlElementRendering,
+} from '../layout/models';
+
+export class LayoutPersonalizationUtils {
+  buildPersonalizedFragment(
+    uid: string,
+    personalizedFragments: { [key: string]: ComponentRendering | null | undefined },
+    defaultComponent: ComponentRendering | null
+  ): ComponentRendering | null {
+    const personalizedFragment = personalizedFragments[uid];
+    if (personalizedFragment === null) {
+      return null;
+    } else if (personalizedFragment === undefined) {
+      return defaultComponent;
+    }
+
+    /**
+     * @param {ComponentRendering} context
+     */
+    function replacePersonalizedRenderings(context: ComponentRendering): boolean {
+      if (context.placeholders) {
+        for (const [key, placeholder] of Object.entries(context.placeholders)) {
+          const hiddenComponents: PersonalizedComponentRendering[] = [];
+          for (let i = 0; i < placeholder.length; i++) {
+            const component = placeholder[i];
+            if (requirePersonalization(component)) {
+              const personalizedFragment = personalizedFragments[component.uid];
+              if (personalizedFragment === null) {
+                hiddenComponents.push(component);
+              } else if (personalizedFragment) {
+                const wasReplaced = replacePersonalizedRenderings(personalizedFragment);
+                if (!wasReplaced) {
+                  return false;
+                }
+                placeholder[i] = personalizedFragment;
+              } else {
+                return false;
+              }
+            }
+          }
+          if (hiddenComponents.length) {
+            context.placeholders[key] = placeholder.filter((component) => {
+              return (
+                !isPersonalizedComponentRendering(component) ||
+                hiddenComponents.indexOf(component) === -1
+              );
+            });
+          }
+        }
+      }
+      return true;
```

Comment:
The replacePersonalizedRenderings function body is difficult to read. Add line spacing. Reduce nesting.